File: Data Structures/Python Data Structures/binary_tree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self):
        self.root = None

    def insert(self, data):
        if self.root == None:
            self.root = Node(data)
        else:
            self._insert(data, self.root)

    def _insert(self, data, current_node):
        if data < current_node.data:
            if current_node.left == None:
                current_node.left = Node(data)
            else:
                self._insert(data, current_node.left)
        else:
            if current_node.right == None:
                current_node.right = Node(data)
            else:
                self._insert(data, current_node.right)


    def print_tree(self):
        if self.root != None:
            self._print_tree(self.root)

    def _print_tree(self, current_node):
        if current_node != None:
            self._print_tree(current_node.left)
            print(str(current_node.data))
            self._print_tree(current_node.right)


    def size_of_tree(self):
        if self.root == None:
            return 0
        else:
            self._size_of_tree(self, current_node)

    def _size_of_tree(self, current_node):
        pass






if __name__ == '__main__':
    btree = BinaryTree()
    btree.insert(3)
    btree.insert(4)
    btree.insert(0)
    btree.insert(8)
    btree.insert(2)
    btree.print_tree()

File: Data Structures/C++ Data Structures/binary_tree.cpp
#include <stdlib.h>
#include <iostream>

using namespace std;

struct node{
  int data;
  node* left;
  node* right;
};

class BinaryTree{
  private:
    void insert(int value, node* currentNode);
    void printTree(node* currentNode);
    node* root; //Pointer called "root" to a node (not a node)
  public:
    BinaryTree(){root = NULL;}
    void insert(int value);
    void printTree();

};

void BinaryTree::insert(int value){
  if (root == NULL){
    //Must point to a node struct called root
    node* head = new node; //Pointer to a node that is a node itself
    head->data = value;
    head->left = NULL;
    head->right = NULL;
    root = head; //head goes out of scope after if-statement
  }else {
    insert(value, root);
  }

}

void BinaryTree::insert(int value, node* currentNode){
  if (value < currentNode->data){
    if (currentNode->left == NULL){
      node* temp = new node;
      temp->data = value;
      temp->left = NULL;
      temp->right = NULL;
      currentNode->left = temp;
    }else{
      insert(value, currentNode);
    }
  }
  else if (value > currentNode->data ){
    if (currentNode->right == NULL){
      node* temp = new node;
      temp->data = value;
      temp->left = NULL;
      temp->right = NULL;
      currentNode->right = temp;
    }else{
      insert(value, currentNode);
    }
  }

}

void BinaryTree::printTree(){
  if(root != NULL){
    printTree(root);
  }
}

void BinaryTree::printTree(node* currentNode){
  if (currentNode != NULL){
    printTree(currentNode->left);
    cout << currentNode->data << endl;
    printTree(currentNode->right);
  }

}

int main(){
  BinaryTree btree;
  btree.insert(0);
  btree.insert(2);
  btree.printTree();

  return 0;
}
